src/execute.c

```diff
@@ -921,6 +1096,243 @@ jv jq_next(jq_state *jq) {
+      }
+      if (jv_get_kind(options) != JV_KIND_OBJECT && jv_get_kind(options) != JV_KIND_NULL) {
+        set_error(jq, jv_invalid_with_msg(jv_string("Eval program options must be an object")));
+        jv_free(program);
+        jv_free(options);
+        jv_free(input);
+        jv_free(args);
+        goto do_backtrack;
+      }
+      if (jv_get_kind(options) == JV_KIND_OBJECT &&
+          jv_object_has(jv_copy(options), jv_string("ALLOW_IO"))) {
+        jv allow_io = jv_object_get(jv_copy(jq->attrs), jv_string("ALLOW_IO"));
+        if (jv_get_kind(allow_io) == JV_KIND_TRUE) {
+          /* Parent allows I/O; child may allow I/O */
+          jv_free(allow_io);
+          options = jv_object_set(jv_copy(jq->attrs), jv_string("ALLOW_IO"),
+                                  jv_object_get(options, jv_string("ALLOW_IO")));
+        } else {
+          /* No I/O for child because no I/O for parent */
+          jv_free(allow_io);
+          jv_free(options);
+          options = jv_copy(jq->attrs);
+        }
+      } else {
+        /* Child will not allow I/O regardless of whether the parent does */
+        jv_free(options);
+        options = jv_object_set(jv_copy(jq->attrs), jv_string("ALLOW_IO"),
+                                jv_false());
+      }
+      if (jv_get_kind(program) != JV_KIND_STRING) {
+        set_error(jq, jv_invalid_with_msg(jv_string("Eval program must be a string")));
+        jv_free(program);
+        jv_free(options);
+        jv_free(input);
+        jv_free(args);
+        goto do_backtrack;
+      }
+
+      jq_state *child = jq_init();
+      if (jv_is_valid(options))
+        jq_set_attrs(jq, jv_copy(options));
+
+      /* Give the child a way to report errors */
+      jv *p = jv_mem_calloc(1, sizeof(jv));
+      *p = jv_null();
+      jq_set_error_cb(child, co_err_cb, p);
+      if (!jq_compile_args(child, jv_string_value(program), jv_copy(args))) {
+        set_error(jq, jv_invalid_with_msg(jv_string_fmt("Eval program failed to compile: %s", jv_string_value(*p))));
+        jq_teardown(&child);
+        jv_free(program);
+        jv_free(options);
+        jv_free(input);
+        jv_free(args);
+        jv_free(*p);
+        free(p);
+        goto do_backtrack;
+      }
+      /* Now let the child send errors to stderr again? */
+      jq_set_error_cb(child, NULL, NULL);
+      jq_start(child, jv_copy(input), jq->debug_trace_enabled ? JQ_DEBUG_TRACE_ALL : 0);
+      copy_callbacks(jq, child);
+      jv_free(program);
+      jv_free(options);
+      jv_free(input);
+      jv_free(args);
+      jv_free(*p);
+      free(p);
+
+      extern struct jq_io_table jq__covt;
+      stack_push(jq, jq_handle_new(jq, "coroutine", &jq__covt, child));
+      stack_save(jq, pc - 1, stack_get_pos(jq));
+      break;
+    }
+    case ON_BACKTRACK(COEVAL): {
+      jv cohandle = stack_pop(jq);
+      jq_handle_close(jq, cohandle);
+      goto do_backtrack;
+    }
+
+    case CORET: {
```

Comment:
why do we need a separate CORET instruction, isn't it possible to distinguish a "top" ret in a child parser context?

Reply:
Ah, because the difference between a top-level `RET` and non-top-level `RET` is that a top-level RET occurs in the context of the top-level frame -- a co-routine's `RET` can't possibly occur in the context of a top-level frame, so the `jq_next()` call on a co-routine's `jq_state *` could never produce an output.  I suppose what I could have done is kept a pointer to the co-routine's "top-level" frame and made `RET` notice that... but I guess I found it easier to add `CORET` -- maybe it's a crutch, but it works well enough.

Reply:
I believe that it should be fairly simple to identify the top frame, since you do that already in stack reset - i.e. you do have some pointer to the topmost co-routine frame.
The reason I think it might be worth doing is that it should save a redundant builtin `cooutput`

Reply:
Yes, but it is a hidden builtin -- you don't get to call it in your jq programs!